sound/soc/sof/ipc4-topology.c

```diff
@@ -1520,6 +1523,13 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 		memcpy(*ipc_config_data + sizeof(*copier_data),
 		       *data, copier_data->gtw_cfg.config_length * 4);
 
+	/* add DMA Config TLV, if configured */
+	if (ipc4_copier->dma_config_tlv.type == SOF_IPC4_GTW_DMA_CONFIG_ID &&
+	    ipc4_copier->dma_config_tlv.length)
```

Comment:
Can we just test `if (ipc4_copier->dma_config_tlv.length)`? Shouldn't we copy dma_config_tlv if ipc4_copier->dma_config_tlv.type is not SOF_IPC4_GTW_DMA_CONFIG_ID?

Reply:
Not following the comment @bardliao 

The two parts that are useful are:
a) the TLV needs to be GTW_DMA_CONFIG, otherwise it's garbage data that cannot be understood by the firmware
b) the length needs to be non-zero.

What would be the merit of testing only one of these indicators?

Reply:
But @plbossart why would we ever set it to anything other than SOF_IPC4_GTW_DMA_CONFIG_ID. This should be set during ipc_setup for the copier case i'd think. And then during prepare, we should add it to the ipc payload if its length is > 0

Reply:
No, I think this type should only be set in the DAI hw_params (or prepare), so that the type and length are NOT modified for pre-LNL platforms. I don't want to touch the TLV in the copier setup, otherwise I need to introduce a platform-specific code to initialize the TLV or not.

So given that the TLV may or may not be initialized by a DAI prepare callback, it seems safe to me to use two indicators. We could downgrade to length only but I don't see what we would be saving.

Reply:
> No, I think this type should only be set in the DAI hw_params (or prepare), so that the type and length are NOT modified for pre-LNL platforms. I don't want to touch the TLV in the copier setup, otherwise I need to introduce a platform-specific code to initialize the TLV or not.
> 
> So given that the TLV may or may not be initialized by a DAI prepare callback, it seems safe to me to use two indicators. We could downgrade to length only but I don't see what we would be saving.

Thanks for the explanation, I got it now. Originally, I thought dma_config_tlv is initial to 0 and dma_config_tlv is always valid if dma_config_tlv.length > 0